### ISSUE TITLE
Stop using CommonJS or AMD dependencies

### ISSUE THREAD

**Issue by vitaly-t** [socketio/socket.io-client#1586]
This library uses `debug` dependency, which prevents any code optimization.

For example, in an Angular web client, we get the following warning in the browser's console:

![image](https://github.com/socketio/socket.io-client/assets/5108906/4fa07c71-9b27-4085-bbff-dd1dcf1ecfaa)

And if we follow [that link](https://angular.io/guide/build#configuring-commonjs-dependencies)...

So basically, we can use that flag to suppress the warnings:

<img width="347" alt="image" src="https://github.com/socketio/socket.io-client/assets/5108906/0a3000f3-51d7-48ac-bef2-174e594a7103">

But hiding the issue isn't really a solution. A library that targets specifically web should NOT include such old CommonJS dependencies anymore ;)

---

* Angular v16.1.2
* Socket.io-client v4.7.0
* Node v18.16.0
* Windows 11

**Comment by darrachequesne** [comment]
Actually, we provide a debug-free ESM build: https://github.com/socketio/socket.io-client/blob/9b235ec01d2ea7f1685b704bd2c5001597635f51/package.json#L26-L30

Not sure why `./build/esm-debug/index.js` is resolved in your case, instead of `./build/esm/index.js`.

**Comment by vitaly-t** [comment]
I'm not really sure what is going on, but it should be very easy to reproduce - just generate a new Angular v16.x project and add this library as a dependency.

b.t.w. If I try to change the import into a specific folder like this:

```ts
import {Manager, Socket} from 'socket.io-client/build/esm';
```

it does not build, comes out with error:
![image](https://github.com/socketio/socket.io-client/assets/5108906/18630218-5105-490a-b15b-1d041f5d745e)

**Comment by vitaly-t** [comment]
I hope [this](https://stackoverflow.com/questions/76545687/socket-io-client-fails-to-provide-an-esm-build-correctly) will be of help.

**Comment by vitaly-t** [comment]
I was able to temporarily solve the issue, by manually changing `./node_modules/socket.io-parser/package.json` file, from this:

```json
"main": "./build/cjs/index.js",
```

to this:

```json
"main": "./build/esm/index.js",
```

**Comment by darrachequesne** [comment]
OK, I think I found the culprit: due to https://github.com/socketio/socket.io-client/commit/781d753a626d01e675056a2ff4e27f5dd599564f (included in v4.7.0), the build that includes the `debug` package (`./build/esm-debug/index.js`) is now imported during development.

Which is OK I guess, since that allows to print the debug logs to the console, and it won't be included in the final bundle (which will use `./build/esm/index.js`). @vitaly-t thoughts?

**Comment by vitaly-t** [comment]
I think, if you have to include that `debug` package, then it should not be the default, or it will confuse developers. I would either expose it through an option, or an explicit import, like `socket.io-client/debug`, just not the default import, because it breaks ESM tree-shaking and causes warnings.

**Comment by darrachequesne** [comment]
> or it will confuse developers.

Totally agree :+1: 

> just not the default import

Just to be sure to be on the same page, it's not really the default import, it's the import when the "development" condition is met:

https://github.com/socketio/socket.io-client/blob/9b235ec01d2ea7f1685b704bd2c5001597635f51/package.json#L26-L30

Reference: https://nodejs.org/api/packages.html#community-conditions-definitions

`import { io } from "socket.io-client/debug"` sounds reasonable.

@FossPrime do you have any thoughts on this?

**Comment by FossPrime** [comment]
> ```ts
> import {Manager, Socket} from 'socket.io-client/build/esm';
> ```
> I guess it is not supposed to be imported that way

Socket.io specifies subpath exports. Which with modern module resolution, prevents access to undeclared endpoints. That same module resolution is causing dev-mode webpack to load the debug package. Tree shacking or production builds should not be affected.

This warning should also be present in Vite's strict modes.

Potential solutions to remove the dev warning:

1. Provide direct access to the builds with a `build/**/*` export, allowing people to override module resolution. I've seen other projects do this successfully.
2. Use another debug package that's ESM friendly, I have not found a good one... So I usually end up writing a mini version of it. 
    - https://github.com/kat-tax/vslite/blob/1de9b5d678719e6b18151fd6409a32769939e3a9/src/utils/debug.ts
3. PR a browser friendly Debug update. Should be much easier than trying to do the whole library like I did (terminal colors, inheritance and module resolution acrobatics make it hard).

If your bundler is asking for a development version with good debuggability and source maps, we should provide it what it's asking for.

Feathers removed the debug dependency for Deno support, and it's caused far worse DX.
This is a growing, common, upstream problem, that should ideally have a common solution.  The people who maintain Debug are active and receptive... But no-one has the endurance to try making a fully backward compatible ESM update.

We should definitely do 1. 3 would be ideal. But 2 is probably more practical in the short term.

If you publish a debug package under @socketio/debug I'll gladly contribute to it and promote it.

**Comment by FossPrime** [comment]
This is worse, for everyone... As now we have to manually turn on and off the debug build and run npm i... Rather than the clean automatic operation we had before.

Even in webpack the biggest downside was an small easy to filter warning in dev.

**Comment by vitaly-t** [comment]
> Deno has no dynamic import support.

[Are you sure? :)](https://medium.com/deno-the-complete-reference/dynamic-imports-in-deno-5e9eb7f66238)

**Comment by darrachequesne** [comment]
> With 4.7.1 downstream libraries that bundle socketio internally have to ALSO make a /debug export for this to work

True, but having people open issues here about the webpack warning was not really thrilling either.

> 3. PR a browser friendly Debug update

That would be ideal indeed.

There is also this fork of `debug`: https://github.com/milahu/debug-esm

Also: https://github.com/vercel/ms/issues/184
